rust/frontend/src/ndarray.rs

```diff
@@ -0,0 +1,359 @@
+//! To copy an NDArray to different context use [`copy_to_ctx`].
+//!
+//! Given a [`Rust's dynamic ndarray`], one can convert it to TVM NDArray as follows:
+//!
+//! # Example
+//!
+//! ```
+//! let a = Array::from_shape_vec((2, 2), vec![1f32, 2., 3., 4.])
+//!     .unwrap()
+//!     .into_dyn(); // Rust's ndarray
+//! let nd = NDArray::from_rust_ndarray(&a, TVMContext::cpu(0), TVMType::from("float")).unwrap();
+//! assert_eq!(nd.shape(), Some(vec![2, 2]));
+//! let rnd: ArrayD<f32> = ArrayD::try_from(&nd).unwrap();
+//! assert!(rnd.all_close(&a, 1e-8f32));
+//! ```
+//!
+//! [`Rust's dynamic ndarray`]:https://docs.rs/ndarray/0.12.1/ndarray/
+//! [`copy_from_buffer`]:struct.NDArray.html#method.copy_from_buffer
+//! [`copy_to_ctx`]:struct.NDArray.html#method.copy_to_ctx
+
+use std::convert::TryFrom;
+use std::mem;
+use std::os::raw::c_int;
+use std::ptr;
+use std::slice;
+
+use num_traits::Num;
+use rust_ndarray::{Array, ArrayD};
+
+use ts;
+
+use Error;
+use Result;
+use TVMByteArray;
+use TVMContext;
+use TVMType;
+
+/// See the [`module-level documentation`](../ndarray/index.html) for more details.
+///
+/// Wrapper around TVM array handle.
+#[derive(Debug)]
+pub struct NDArray {
+    pub(crate) handle: ts::TVMArrayHandle,
+    is_view: bool,
+}
+
+impl NDArray {
+    pub(crate) fn new(handle: ts::TVMArrayHandle, is_view: bool) -> Self {
+        NDArray {
+            handle: handle,
+            is_view: is_view,
+        }
+    }
+
+    /// Returns the underlying array handle.
+    #[inline]
+    pub fn as_handle(&self) -> ts::TVMArrayHandle {
+        self.handle
+    }
+
+    #[inline]
+    pub fn is_view(&self) -> bool {
+        self.is_view
+    }
+
+    /// Returns the shape of the NDArray.
+    pub fn shape(&self) -> Option<Vec<usize>> {
+        let arr = unsafe { *(self.handle) };
+        if arr.shape.is_null() || arr.data.is_null() {
+            return None;
+        };
+        let slc = unsafe { slice::from_raw_parts_mut(arr.shape as *mut usize, arr.ndim as usize) };
+        Some(slc.to_vec())
+    }
+
+    /// Returns the total number of entries of the NDArray.
+    #[inline]
+    pub fn size(&self) -> Option<usize> {
+        self.shape().map(|v| v.into_iter().product())
+    }
+
+    /// Returns the context which the NDArray was defined.
+    pub fn ctx(&self) -> TVMContext {
+        let dlctx = unsafe { (*self.handle).ctx };
+        dlctx.into()
+    }
+
+    #[inline]
+    pub fn context(&self) -> TVMContext {
+        self.ctx()
+    }
+
+    /// Returns the type of the entries of the NDArray.
+    pub fn dtype(&self) -> TVMType {
+        let dldtype = unsafe { (*self.handle).dtype };
+        dldtype.into()
+    }
+
+    /// Returns the number of dimensions of the NDArray.
+    pub fn ndim(&self) -> usize {
+        unsafe { (*self.handle).ndim as usize }
+    }
+
+    pub fn strides(&self) -> Option<Vec<usize>> {
+        unsafe {
+            (*self.handle).strides.as_mut().map(|pv| {
```

Comment:
woah, `unsafe` and `mut`? much safer would be to convert `strides` to a slice and then `to_vec` that (but do you _really_ need to return an owned array?)